[SPARK-47764][FOLLOW-UP] Change to use ShuffleDriverComponents.removeShuffle to remove shuffle properly

bozhang2820 opened the pull request:
### What changes were proposed in this pull request?
This is a follow-up for https://github.com/apache/spark/pull/45930, where we introduced ShuffleCleanupMode and implemented cleaning up of shuffle dependencies. 

There was a bug where `ShuffleManager.unregisterShuffle` was used on Driver, and in non-local mode it is not effective at all. This change fixed the bug by changing to use `ShuffleDriverComponents.removeShuffle` instead.


### Why are the changes needed?
This is to address the comments in https://github.com/apache/spark/pull/45930#discussion_r1584223064

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Updated unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No

bozhang2820 committed:
change to use shuffleDriverComponents.removeShuffle

dongjoon-hyun commented:
Could you re-trigger CI, @bozhang2820 ?

dongjoon-hyun commented:
I converted this to `Draft` PR because the CI is broken and the PR title has `[WIP]`.

bozhang2820 committed:
Fix cleanup in test

bozhang2820 committed:
Cleanup based on isLocal

bozhang2820 commented:
> Could you re-trigger CI, @bozhang2820 ?

Sorry for the late reply @dongjoon-hyun.. Updated the test and removed the WIP tag. 

It seems that the current test failure is unrelated to this change?

bozhang2820 committed:
Trigger CI

abellina commented:
@bozhang2820 this PR is needed for the feature to work in my tests, I am just wondering what the status of this is.

mridulm reviewed (commented):
Looks reasonable - can you add a test for this ?
We can use `local` vs `local-cluster` to test the expected behavior

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -176,7 +176,13 @@ object SQLExecution extends Logging {
+                  if (sc.isLocal) {
+                    SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  } else {
+                    sc.shuffleDriverComponents.removeShuffle(shuffleId, false)
```
does this work for local mode?

bozhang2820 replied:
Yes `ShuffleDriverComponents.removeShuffle` should work for both local an non-local mode. Will update the code here.

bozhang2820 commented:
> Looks reasonable - can you add a test for this ? We can use `local` vs `local-cluster` to test the expected behavior

I tried adding a unit test for local-cluster mode but found it a bit difficult:
1. The cleanup is done in a best-effort basis, and in local-cluster mode the shuffle blocks are less likely to be cleaned up in time. 
2. There seems to be no API to get the MigratableResolver on executors. 

What do you think?

bozhang2820 committed:
remove if isLocal

bozhang2820 committed:
Merge remote-tracking branch 'apache/master' into spark-47764-1

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -177,7 +177,9 @@ object SQLExecution extends Logging {
-                  SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  // Do not unregister the shuffle on MapOutputTracker here to trigger stage retry.
+                  // Otherwise, downstream tasks will fail with MetadataFetchFailedException.
+                  sc.shuffleDriverComponents.removeShuffle(shuffleId, Utils.isTesting)
```
This cleanup is not critical, I think the second parameter `blocking` can always be false

bozhang2820 replied:
The unit tests would be flaky if we use false here.

cloud-fan replied:
oh I see, can we add some comments to explain it?

bozhang2820 replied:
Sure. Done.

bozhang2820 committed:
Update comments

dongjoon-hyun commented:
Thank you, @bozhang2820, @cloud-fan , @mridulm .
Merged to master for Apache Spark 4.0.0-preview2.

dongjoon-hyun commented:
For the test cases, please file a JIRA issue not to forget because it seems to need more time, @bozhang2820 .
